api/controller/assignments.js

```diff
@@ -0,0 +1,321 @@
+            if (err) {
+                res.status(500).send({
+                    error: "Error accessing database",
+                    code: 23
+                });
+                return;
+            }
+            if (item) {
+                res.status(400).send({
+                    error: "Assignment with this name already exists",
+                    code: 24
+                });
+                return;
+            }
+            collection.insertOne(assignment, { safe: true }, function (err, result) {
+                if (err) {
+                    res.status(500).send({
+                        error: "Error accessing database",
+                        code: 23
+                    });
+                    return;
+                }
+                res.send(result.ops[0]);
+            }
+            );
+        });
+    }
+    );
+};
+
+//find All assignments 
+exports.findAll = function (req, res) {
+    //find all assignments 
+    db.collection(assignmentCollection, function (err, collection) {
+        if (err) {
+            res.send({ 'error': 'An error has occurred in finding assignments' });
+        } else {
+            collection.find().toArray(function (err, items) {
+                res.send(items);
+
+            });
+        }
+    });
+};
+
+exports.launchAssignment = function (req, res) {
+    //validate
+    if (!req.params.assingid) {
+        res.status(400).send({
+            error: "Assignment id is not defined",
+            code: 22
+        });
+        return;
+    }
+
+    //find assignment by id
+    db.collection(assignmentCollection, function (err, collection) {
+        collection.findOne({ _id: new mongo.ObjectID(req.params.assingid) }, function (err, assignment) {
+            if (err) {
+                res.send({ 'error': 'An error has occurred in finding assignment' });
+            } else {
+                var assignmentData;
+                assignmentData = assignment;
+                //find all students from classrooms
+                var classrooms = assignmentData.classrooms;
+                var privateJournalIds = []; //to store private journal ids of every student
+                var students = []; //to store all students data 
+
+
+                //get students private_journal
+                common.fetchAllStudents(classrooms).then(function (stds) {
+                    //student data
+                    students = students.concat(stds);
+
+                    //getting priavte journal ids from every student id
+                    var uniqueJournalIds = [];
+                    for (var i = 0; i < stds.length; i++) {
+                        if (uniqueJournalIds.indexOf(stds[i].private_journal.toString()) == -1) {
+                            uniqueJournalIds.push(stds[i].private_journal.toString());
+                        }
+                    }
+
+                    privateJournalIds = uniqueJournalIds;
+                    var entryDoc = {};
```

Comment:
This variable does not needs to be initiated so sooner, it can be initiated where it's defined for the first time.